lib/streamlit/web/server/memory_media_file_storage.py

```diff
@@ -0,0 +1,156 @@
+import hashlib
+import mimetypes
+from typing import Union, NamedTuple, Dict, Optional, List
+
+from streamlit.logger import get_logger
+from streamlit.runtime.media_file_storage import MediaFileStorage, MediaFileStorageError
+from streamlit.runtime.stats import CacheStatsProvider, CacheStat
+
+LOGGER = get_logger(__name__)
+
+PREFERRED_MIMETYPE_EXTENSION_MAP = {
+    "image/jpeg": ".jpeg",
+    "audio/wav": ".wav",
+}
+
+
+def _calculate_file_id(
+    data: bytes, mimetype: str, filename: Optional[str] = None
+) -> str:
+    """Hash data, mimetype, and an optional filename to generate a stable file ID.
+
+    Parameters
+    ----------
+    data
+        Content of in-memory file in bytes. Other types will throw TypeError.
+    mimetype
+        Any string. Will be converted to bytes and used to compute a hash.
+    filename
+        Any string. Will be converted to bytes and used to compute a hash.
+    """
+    filehash = hashlib.new("sha224")
+    filehash.update(data)
+    filehash.update(bytes(mimetype.encode()))
+
+    if filename is not None:
+        filehash.update(bytes(filename.encode()))
+
+    return filehash.hexdigest()
+
+
+def _get_extension_for_mimetype(mimetype: str) -> str:
+    # Python mimetypes preference was changed in Python versions, so we specify
+    # a preference first and let Python's mimetypes library guess the rest.
+    # See https://bugs.python.org/issue4963
+    #
+    # Note: Removing Python 3.6 support would likely eliminate this code
+    if mimetype in PREFERRED_MIMETYPE_EXTENSION_MAP:
+        return PREFERRED_MIMETYPE_EXTENSION_MAP[mimetype]
+
+    extension = mimetypes.guess_extension(mimetype)
+    if extension is None:
+        return ""
+
+    return extension
+
+
+class MemoryFile(NamedTuple):
+    content: bytes
+    mimetype: str
+    filename: Optional[str]
+
+
+class MemoryMediaFileStorage(MediaFileStorage, CacheStatsProvider):
+    def __init__(self, media_endpoint: str):
+        self._files_by_id: Dict[str, MemoryFile] = {}
+        self._media_endpoint = media_endpoint
+
+    def load_and_get_id(
+        self,
+        path_or_data: Union[str, bytes],
+        mimetype: str,
+        filename: Optional[str] = None,
+    ) -> str:
+        """Add a file to the manager and return its ID."""
+        file_data: bytes
+        if isinstance(path_or_data, str):
+            file_data = self._read_file(path_or_data)
+        else:
+            file_data = path_or_data
+
+        # Because our file_ids are stable, if we already have a file with the
+        # given ID, we don't need to create a new one.
+        file_id = _calculate_file_id(file_data, mimetype, filename)
+        media_file = self._files_by_id.get(file_id)
+        if media_file is None:
+            LOGGER.debug("Adding media file %s", file_id)
+            media_file = MemoryFile(
+                content=file_data, mimetype=mimetype, filename=filename
+            )
+            self._files_by_id[file_id] = media_file
```

Comment:
super nit: I think this would be cleaner as something like

```python
if file_id not in self._files_by_id:
    LOGGER.debug("Adding media file %s", file_id)
    self._files_by_id[file_id] = MemoryFile(
        content=file_data, mimetype=mimetype, filename=filename
    )
```
since the `media_file` variable is really only used for an `is None` check.